packaging/greenboot/microshift-greenboot-required.sh

```diff
@@ -0,0 +1,194 @@
+
+# Source Greenboot configuration file if it exists
+GREENBOOT_CONF_FILE=/etc/greenboot/greenboot.conf
+[ -f "${GREENBOOT_CONF_FILE}" ] && source ${GREENBOOT_CONF_FILE}
+WAIT_TIMEOUT_SECS_BASE=${MICROSHIFT_WAIT_TIMEOUT_SEC:-300}
+
+# Check the current user before the output redirect
+if [ $(id -u) -ne 0 ] ; then
+    echo "The '${SCRIPT_NAME}' script must be run with the 'root' user privileges"
+    exit 0
+fi
+
+# Set the exit handler to log the exit status
+trap 'script_exit' EXIT
+
+# The script exit handler logging the FAILURE or FINISHED message depending
+# on the exit status of the last command
+#
+# args: None
+# return: None
+function script_exit() {
+    [ "$?" -ne 0 ] && status=FAILURE || status=FINISHED
+    echo $status
+}
+
+# Run a command every second until it returns a zero exit status
+#
+# arg1: Time in seconds to wait for a command to succeed
+# argN: Command to run with optional arguments
+# return: 0 if a command ran successfully within the wait period, or 1 otherwise
+function wait_for() {
+    local timeout=$1
+    shift 1
+
+    until [ $timeout -le 0 ] || ("$@"); do
+        sleep 1
+        timeout=$(( timeout - 1 ))
+    done
+
+    [ $timeout -le 0 ] && return 1
+    return 0
+}
+
+# Check the microshift.service systemd unit activity, terminating the script
+# with the SIGTERM signal if the unit reports a failed state
+#
+# args: None
+# return: 0 if the systemd unit is active, or 1 otherwise
+function microshift_service_active() {
+    local is_failed=$(systemctl is-failed microshift.service)
+    local is_active=$(systemctl is-active microshift.service)
+
+    # Terminate the script in case of a failed service - nothing to wait for
+    if [ "${is_failed}" = "failed" ] ; then
+        echo "The microshift.service systemd unit is failed. Terminating..."
+        kill -TERM ${SCRIPT_PID}
+    fi
+    # Check the service activity
+    [ "${is_active}" = "active" ] && return 0
+    return 1
+}
+
+# Check if any MicroShift pods are in the 'Running' status
+# 
+# args: None
+# return: 0 if any pods are in the 'Running' status, or 1 otherwise
+function any_pods_running() {
+    local count=$(oc get pods ${OCGET_OPT} -A 2>/dev/null | awk '$4~/Running/' | wc -l)
+
+    [ "${count}" -gt 0 ] && return 0
+    return 1
+}
+
+# Check if a given number of MicroShift pods in a given namespace are in the 'Ready' status,
+# terminating the script with the SIGTERM signal if more pods are ready than expected.
+#
+# args: None
+# env1: 'CHECK_PODS_NS' environment variable for the namespace to check
+# env2: 'CHECK_PODS_CT' environment variable for the pod count to check
+# return: 0 if the expected number of pods are ready, or 1 otherwise
+function namespace_pods_ready() {
+    local ns=${CHECK_PODS_NS}
+    local ct=${CHECK_PODS_CT}
+
+    local status=$(oc get pods ${OCGET_OPT} -n ${ns} -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}' 2>/dev/null)
```

Comment:
Something to keep in mind is that we are adding a dependency to the openshift-client with this patch. I am not sure how desirable it is on all situations, but also I am not sure how could we get away without it. One possibility could be to have microshift report the running status via a simple localhost API endpoint, or a file/unixsocket.




I

Reply:
Created the USHIFT-772 issue as a follow-up.